File: src/MergeTwoSortedLists.java
/**
 * User: guoxc
 * Date: 15-1-9
 * Time: 19:13
 */
public class MergeTwoSortedLists {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode pt1 = l1;
        ListNode pt2 = l2;
        ListNode head = null;
        if (pt1 == null && pt2 == null) {
            return null;
        } else if (pt1 == null) {
            head = new ListNode(pt2.val);
            pt2 = pt2.next;
            copyLinkList(pt2, head);
            return head;
        } else if (pt2 == null) {
            head = new ListNode(pt1.val);
            pt1 = pt1.next;
            copyLinkList(pt1, head);
            return head;
        }
        if (pt1.val < pt2.val) {
            head = new ListNode(pt1.val);
            pt1 = pt1.next;
        } else {
            head = new ListNode(pt2.val);
            pt2 = pt2.next;
        }
        ListNode x = head;
        while (pt1 != null && pt2 != null) {
            if (pt1.val < pt2.val) {
                x.next = new ListNode(pt1.val);
                x = x.next;
                pt1 = pt1.next;
            } else {
                x.next = new ListNode(pt2.val);
                x = x.next;
                pt2 = pt2.next;
            }
        }
        copyLinkList(pt1, x);
        copyLinkList(pt2, x);
        return head;
    }
    public void copyLinkList(ListNode pt, ListNode x){
        while(pt != null){
            x.next = new ListNode(pt.val);
            x = x.next;
            pt = pt.next;
        }
    }
    public static void test(){
        ListNode l1 = new ListNode(new int[]{1,2,4,6,10});
        ListNode l2 = new ListNode(new int[]{3,4,5,6,7});
        System.out.println(new MergeTwoSortedLists().mergeTwoLists(l1,l2));
    }
}

File: cpp/merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* res = new ListNode(0);
        ListNode* p = res;
        ListNode* p1 = l1;
        ListNode* p2 = l2;
        while(p1 != NULL && p2 != NULL){
            if(p1->val < p2->val){
                p->next = p1;
                p1 = p1->next;
                p = p->next;
                p->next = NULL;
            }else{
                p->next = p2;
                p2 = p2->next;
                p = p->next;
                p->next = NULL;
            }
        }
        if(p1 != NULL){
            p->next = p1;
        }else if(p2 != NULL){
            p->next = p2;
        }
        return res->next;

    }
};
